posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,172 @@
+
+Now it could be very common to have a huge amount of `Item`s in our DB that a `List` will not fit into memory and / or it will be a very expensive operation. So we might want to stream the results of `findAll` instead of have them all in memory on a `List`, making `Doobie` a great candidate for the job. But wait... We have a problem now. Our `ItemRepository` algebra has fixed the definition of `findAll` as `F[List[Item]]` so we won't be able to create an interpreter that returns a streaming result instead.
+
+### Rethinking our algebra
+
+We should think about abstracting over that `List` and two of the most common abstractions that immediately come to mind are `Foldable` and `Traverse`. But although these typeclasses are very useful, they are not enough to represent a stream of values, so we should come up with a better abstraction.
+
+Well, it seems that our options are either adding another higher-kinded parameter `G[_]` to our algebra or just define an abstract member `G[_]`. So let's go with the first one:
+
+```scala
+trait ItemRepository[F[_], G[_]] {
+  def findAll: G[Item]
+  def find(name: ItemName): F[Option[Item]]
+  def save(item: Item): F[Unit]
+  def remove(name: ItemName): F[Unit]
+}
+```
+
+Great! This looks good so far.
+
+### Streaming support interpreter
+
+Now let's write a new `PostgreSQL` interpreter with streaming support:
+
+```scala
+import fs2.Stream
+
+// Pretend this is the real implementation... What matters for the example are the types.
+class StreamingItemRepository[F[_]](implicit F: Sync[F]) extends ItemRepository[F, Stream[F, ?]] {
+  private val mockItem = Item(ItemName("laptop"), BigDecimal(1500))
+
+  override def findAll: Stream[F, Item] = Stream.eval(F.pure(mockItem))
+  override def find(name: ItemName): F[Option[Item]] = F.pure(None)
+  override def save(item: Item): F[Unit] = F.unit
+  override def remove(name: ItemName): F[Unit] = F.unit
+}
+
+```
+
+Voilà! We got our streaming implementation of `findAll`.
+
+### Test interpreter
+
+That's all we wanted, but what about testing it? Sure, we might prefer to have a simple implementation by just using a plain `List`, so what can we possibly do?
+
+```scala
+object MemRepository extends ItemRepository[Id, List] {
+  private val mem = MutableMap.empty[String, Item]
+
+  override def findAll: List[Item] = mem.headOption.map(_._2).toList
+  override def find(name: ItemName): Id[Option[Item]] = mem.get(name.value)
+  override def save(item: Item): Id[Unit] = mem.update(item.name.value, item)
+  override def remove(name: ItemName): Id[Unit] = {
+    mem.remove(name.value)
+    ()
+  }
+}
+```
+
+That's pretty much it! We managed to abstract over the return type of `findAll` by just adding an extra parameter to our algebra.
```

Comment:
This is nice and all, but I don't see how the pieces would compose together. Say, I want to write a generic function which takes all items (using `findAll`), applies some discounts (I'm assuming the second parameter of `Item` is a price) and writes them back using `save`. If I specified `Stream` or `List`, it'd be immediately obvious how to do that, but not with that second type parameter which I cannot quite constraint. Such an example would be very interesting.

Reply:
Fair point. I actually have a bigger example putting all the pieces together and showing that to integrate the streaming `findAll` with `save` you'll need to compile that streaming to an `F` using `compile`, will try to fit it in the post whilst trying to keep it concise.

Reply:
I just re-read your comment (missed the "generic function" part before). I'll try to add a short section of why I wouldn't recommend this and strictly do it at the interpreter level instead.